internal/pkg/runtime/engines/singularity/process_linux.go

```diff
@@ -193,6 +193,9 @@ func (engine *EngineOperations) StartProcess(masterConn net.Conn) error {
 
 	if (!isInstance && !shimProcess) || bootInstance || engine.EngineConfig.GetInstanceJoin() {
 		err := syscall.Exec(args[0], args, env)
+		if _, errNoShell := os.Stat("/bin/sh"); os.IsNotExist(errNoShell) {
```

Comment:
As I mentioned in the issue, I think this is happening too late. We should check for `/bin/sh` much earlier (when we decide which wrapper to launch, in `checkExec`) so that we can provide a better error message there?

Reply:
Resolved, I think. Thanks!